solver/src/naive_amm_settlement.rs

```diff
@@ -0,0 +1,301 @@
+    // where [price_a, price_b] is the price vector. These prices can only exist if:
+    // sell_a.buy_amount * sell_b.buy_amount <= sell_a.sell_amount * sell_b.sell_amount
+    // This transformation avoids lossy division.
+    let left = sell_a.buy_amount.checked_mul(sell_b.buy_amount)?;
+    let right = sell_a.sell_amount.checked_mul(sell_b.sell_amount)?;
+    let have_price_overlap = left <= right;
+    if !have_price_overlap {
+        return None;
+    }
+    amm_match_(sell_a, sell_b)
+}
+
+// Match two orders directly with their full sell amounts.
+fn direct_match(sell_a: &UserOrder, sell_b: &UserOrder) -> Settlement {
+    Settlement {
+        clearing_prices: [
+            (sell_a.sell_token, sell_a.sell_amount),
+            (sell_b.sell_token, sell_b.sell_amount),
+        ]
+        .iter()
+        .copied()
+        .collect(),
+        trades: vec![
+            Trade {
+                order: *sell_a,
+                executed_amount: sell_a.sell_amount,
+            },
+            Trade {
+                order: *sell_b,
+                executed_amount: sell_b.sell_amount,
+            },
+        ],
+        ..Default::default()
+    }
+}
+
+// Match two orders with amm assuming that there is price overlap.
+fn amm_match_(sell_a: &UserOrder, sell_b: &UserOrder) -> Option<Settlement> {
+    // Based on our assumptions we know that exactly one order is "bigger" than the other in the
+    // sense that it a larger sell amount than the other order's buy amount.
+    // It is not possible for both orders to be bigger because that would be a direct match which
+    // has already been handled.
+    let (big, small) = if sell_a.sell_amount > sell_b.buy_amount {
+        (sell_a, sell_b)
+    } else {
+        (sell_b, sell_a)
+    };
+    // Unwrap because of the above explanation.
+    let bigger_sing_buy_amount = big.buy_amount.checked_sub(small.sell_amount).unwrap();
+    // Because the smart contract enforces uniform prices we must pick one price that is accepted by
+    // both orders. We know that there is price overlap so we could pick any price between the price
+    // limits of the orders. We pick the price of the bigger order because it's sell token surplus
+    // will have to be traded with the amm and the bigger order always has a more lax amm price
+    // requirement of the bigger order's sell token than the smaller order.
+    // Because we picked the bigger order's price we must add an extra buy amount to the smaller
+    // order so that it gets the same price.
+    let small_buy_amount = rounded_up_division(
+        small.sell_amount.checked_mul(big.sell_amount)?,
+        big.buy_amount,
+    )?;
+    // Does this have to be saturating or is guaranteed to not underflow? With the rounding from the
+    // previous operation I suspect it might.
+    let big_extra_sell_amount = big.sell_amount.saturating_sub(small_buy_amount);
+    if big_extra_sell_amount.is_zero() {
```

Comment:
My intuition is that with exact math it cannot happen because if the extra sell amount is zero and there is price overlap it must be a direct match, yes. But in this case we have the mentioned rounding error so I am not certain.